File: java_generator/Reader4.java
public class Reader4 {
    private final int BUF_SZ = 4;
    private int idx;
    private final char[] data;

    Reader4(String s) {
        idx = 0;
        data = s.toCharArray();
    }

    int read4(char[] buf) {
        int bytes = Math.min(data.length - idx, BUF_SZ);
        System.arraycopy(data, idx, buf, 0, bytes);
        idx += bytes;
        return bytes;
    }
}

File: _designs_001/Read N Characters Given Read4 II/Reader4.rb
class Reader4
  def initialize(s)
    @data = s
    @buffer_size = 4
    @idx = 0
  end
  
  def read(buf)
    bytes = [@data.length - @idx, @buffer_size].min
    buf.delete!(buf)
    buf << @data[@idx...@idx + bytes]
    @idx += bytes
    bytes
  end
end
